미션 제출합니다.

Gongmeda opened the pull request.

Gongmeda committed:
docs: 구현할 기능 목록 정리

Gongmeda committed:
feat: Car & CarMoveRule 구현

Gongmeda committed:
docs: 구현할 기능 목록 수정

Commit:
feat: Car 이름 validation 구현

Commit:
feat: carRace 진행 구현

Gongmeda committed:
feat: CarRace.getWinningCars 구현 및 필요 없는 메소드 제거

Commit:
feat: 자동차 경주 Controller, View, Application 구현

Gongmeda committed:
refactor: 클래스 분리 및 재구성

- 패키지 구조 정리
- NumberGenerator로 랜덤 숫자 생성 로직 분리 및 구현
- 인덴트 규칙에 맞게 메소드 분리

Commit:
feat: CarRaceView에서 입력에 대한 유효성 검사

Gongmeda committed:
refactor: 테스트 코드 중복 제거 및 정리

Commit:
chore: 테스트 코드 패키지 이동

Commit:
feat: String calculator 기능 구현

Gongmeda committed:
refactor: StringCalculator 코드 리팩토링

liquidjoo commented on src/main/java/carrace/CarRaceApplication.java:
```diff
@@ -0,0 +1,10 @@
+
+public class CarRaceApplication {
+
+	public static void main(String[] args) {
```
MVC 패턴에 대해서는 다음 미션에 더욱 자세히 리뷰하도록 하겠습니다~

liquidjoo commented on src/main/java/carrace/CarRaceController.java:
```diff
@@ -0,0 +1,53 @@
+		return view.getCarRaceRound();
+	}
+
+	private void proceedRounds(CarRace carRace, int round) {
+		view.displayResultStartMessage();
+		view.displayRoundResult(carRace.getCars());
+		for (int r = 0; r < round; r++) {
+			carRace.runRound();
+			view.displayRoundResult(carRace.getCars());
+		}
+	}
```
View 와 비즈니스 로직을 분리해 볼 수 있을까요?

Gongmeda replied:
조금 이해가 잘 안되는 것 같습니다! 이미 view와 비즈니스로직이 각각 `CarRaceView`와 `CarRace`로 나눠져 구현되었다고 생각하는데요. `round` 횟수만큼 경주를 진행하는 것(for loop)가 controller에 구현되어 있는것에 대한 말씀이실까요? 조금 더 설명 부탁드립니다!

Gongmeda replied:
round 횟수만큼 경주를 한다는 부분을 CarRace 쪽으로 넘겨서 처리했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/1967b0fcbe34d4d344df93d662f849351f9f0c0e

liquidjoo commented on src/main/java/carrace/CarRaceView.java:
```diff
@@ -0,0 +1,70 @@
+		}
+	}
+
+	public void validateDuplicateNames(List<String> carNames) {
+		Set<String> carNamesSet = new HashSet<>(carNames);
+		if (carNames.size() != carNamesSet.size()) {
+			throw new IllegalArgumentException("중복된 이름이 존재합니다.");
+		}
+	}
```
해당 로직은 View 의 로직 보단 비즈니스 로직에 가까워 보이네요 :)
별도의 역할을 만들어 책임을 이동 시켜보면 어떨까요?

Gongmeda replied:
처리했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/d335afd98b2fdb97dee0c5d7ca9c7183404fe4ad

Gongmeda replied:
`CarRace` 클래스로 옮겼습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/1c5076eafe1d0d96de6372804732aaf1eb33976c

liquidjoo commented on src/main/java/carrace/CarRaceView.java:
```diff
@@ -0,0 +1,70 @@
+		return round;
+	}
+
+	public void validateRaceRound(int round) {
+		if (round <= 0) {
+			throw new IllegalArgumentException("라운드는 양수여야 합니다.");
+		}
+	}
```
해당 로직은 View 의 로직 보단 비즈니스 로직에 가까워 보이네요 :)
별도의 역할을 만들어 책임을 이동 시켜보면 어떨까요?

Gongmeda replied:
처리했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/d335afd98b2fdb97dee0c5d7ca9c7183404fe4ad

Gongmeda replied:
`Round` 라는 클래스를 만들고, 거기서 해당 validation을 처리하라는 말씀이신가요?

Gongmeda replied:
반영했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/e053e58c57024b3a1e2d3e19208b5e5fa070586a

liquidjoo commented on src/main/java/carrace/domain/Car.java:
```diff
@@ -0,0 +1,23 @@
+        return position;
+    }
+
+    public void moveForward() {
+        position++;
+    }
```
void 타입도 좋지만 반환 타입을 가져보는 것은 어떨까요?
어떤 장점이 있을까요?

Gongmeda replied:
생각나는 방식은
1. `int` 타입으로 새롭게 변경된 position의 값을 반환
2. `Car` 타입으로 자기자신을 반환하여 메소드 체이닝을 가능하게 하는 것
두 가지가 생각나는데요.
이 두 방식에서 오는 장점이 현재 구현된 코드상에서는 테스트 코드에서 생기는 장점밖에 없다고 생각이 됩니다. 혹시 다른 방식을 생각하시고 말씀해주신걸까요?

liquidjoo replied:
테스트 코드의 장점도 있으나 moveFoward 후 한번 더 해당 해당 상태 값을 가져와 비교해야하는 점에서 직관적으로 표현하면 코드 한 줄이라도 줄여볼 수 있을 것 같네요 :)

Gongmeda replied:
반영했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/1967b0fcbe34d4d344df93d662f849351f9f0c0e

liquidjoo commented on src/main/java/carrace/domain/CarMoveRule.java:
```diff
@@ -0,0 +1,16 @@
+
+    private static final int MOVE_THRESHOLD = 4;
+
+    private final NumberGenerator numberGenerator;
+
+    public CarMoveRule(NumberGenerator numberGenerator) {
+        this.numberGenerator = numberGenerator;
+    }
```
전략패턴 👍 
이러한 전략패턴을 사용할 때 어떤 장점이 있을까요?

Gongmeda replied:
숫자 생성 로직을 유연하게 변경할 수 있는 장점이 있는 것 같습니다! 따라서 숫자 생성 로직이 랜덤이 아니게 되어도 `CarMoveRule` 코드를 변경하지 않고도 처리할 수 있을 것 같습니다.
추가로, 테스트 코드를 작성할 때도 원하는 케이스를 쉽게 세팅할 수 있게도 하는 것 같습니다!

liquidjoo commented on src/main/java/carrace/domain/CarRace.java:
```diff
@@ -0,0 +1,36 @@
+		this.cars = cars;
+	}
+
+	public void runRound() {
+		for (Car car : cars) {
+			moveIfMovable(car);
+		}
+	}
+
+	private void moveIfMovable(Car car) {
+		if (carMoveRule.isMovable()) {
+			car.moveForward();
+		}
+	}
```
runRound 를 수행 후 View 를 보여주는 것 보다 모든 라운드를 수행 후 결과를 View 에 넘겨주면 어떨까요?

Gongmeda replied:
round를 전부 실행한 후 View에 넘겨주기 위해서는 N회의 round를 진행한다 가정했을 때 각 round의 진행 결과의 collection을 `CarRace` 측에서 생성하여 반환해야 할 것 같은데요, 현재의 구현에서 변경이 좀 많이 필요할 것 같다는 생각이 듭니다.
현재의 요구 사항에서 말씀하신대로 바꿨을 때의 장점이 뭔지 크게 와닿지 않는 것 같은데 설명해주실 수 있으실까요?

liquidjoo commented on src/main/java/carrace/domain/CarRace.java:
```diff
@@ -0,0 +1,36 @@
+		}
+	}
+
+	public List<Car> getWinningCars() {
+		int maxPosition = cars.stream().mapToInt(Car::getPosition).max().orElse(0);
+		return cars.stream().filter(car -> car.getPosition() == maxPosition).collect(Collectors.toList());
+	}
```
별도의 역할이 나와도 좋을 것 같아요~ 어떤 역할이 나올 수 있을까요?

Gongmeda replied:
이 부분에서 나올 수 있는 역할로 `Car`의 목록을 나타내는 클래스인 `Cars`를 구현할 수 있을 것 같습니다.
`Cars.getMaxCarPosition`와 `Cars.getCarsOfPosition(position)` 메소드를 구현해서 `getWinningCars`의 두 줄을 각각 대체할 수 있을 것 같은데요. 저는 '우승자를 고르는 기준이 변경된다면?'이라는 관점에서 `Cars`를 구현하는게 불필요한 중간 단계를 만드는게 아닌가 하는 의문이 조금 드네요. 이에 대해 의견 여쭙고 싶습니다!

liquidjoo replied:
고르는 기준은 별도의 객체에서 반환해주면 될 것 같아요~
이길 때의 값은 모든 실행이 끝난 이후인 Cars 객체에서 필요한 정보만 넘겨주면 될 것 같아 보이네요 :)

liquidjoo commented on src/main/java/carrace/domain/CarRace.java:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	public List<Car> getCars() {
+		return this.cars;
```
방어적 복사를 활용해 보면 좋을 것 같아요~
이번 기회에 방어적 복사에 대해서 알아보아요

Gongmeda replied:
반영했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/64429dd077b685a8af43628eb183db9abafcaed4

liquidjoo commented on src/main/java/carrace/domain/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,11 @@
+
+    @Override
+    public int generate() {
+        return (new Random()).nextInt(10);
```
Random 객체가 생성되는 비용을 어떻게 줄여볼 수 있을까요?

Gongmeda replied:
static 변수로 생성 비용을 절감해봤습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/28f00a242a9052b084e7b9231308022f54371d6b

liquidjoo commented on src/main/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringCalculator {
```
상태값 없이 메서드만 존재하는 객체를 불완전 객체라고 불리는데요
수업시간에 진행한 것에서 한번 더 리팩토링 하는 것으로
이를 완전한 객체로 만들어 보면 어떨까요?

또한 너무 많은 책임이 있는 것 같아요
책임과 역할을 잘나누어 보면 좋을 것 같아요 :)

Gongmeda replied:
문자열 파싱에 대한 책임은 따로 분리하는게 맞다 판단하여 처리했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/627721ef157852bfb75db2e4d1d0aa96e7207c1b
완전 객체로 변경하는 부분은 어떤 상태값을 가지게 해야할지 모르겠어서 여쭤보고 싶습니다!

liquidjoo replied:
List<Integer> 로 상태 값을 가져보면 될 것 같은데요 :)

liquidjoo commented on src/main/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+
+    public static int calculate(String input) {
+        if (isNullOrBlank(input)) {
+            return 0;
```
이러한 숫자, 문자를 상수로 표현해 명확히 표현하면 좋을 것 같아요 :)

Gongmeda replied:
반영했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/a3e0e60090db91cbf0b6264a193e3373cc385e88

liquidjoo commented on src/main/java/stringcalculator/StringCalculator.java:
```diff
@@ -0,0 +1,44 @@
+            return 0;
+        }
+
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(input);
```
Pattern의 complie 는 생성 비용이 많이 들어요
이를 어떻게 개선해 볼 수 있을까요?

Gongmeda replied:
static 변수로 `Pattern.compile`을 미리 실행하고 재사용하여 비용을 절감해봤습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/c9a14ada2c79add578b942c8418991aa69d2d902

liquidjoo commented on src/test/java/carrace/CarRaceTest.java:
```diff
@@ -0,0 +1,88 @@
+        cars = List.of(car1, car2, car3);
+    }
+
+    @Test
```
`@DisplayName()` 을 활용해 한글로도 테스트 코드의 이름을 표현해보면 좋습니다

Gongmeda replied:
반영했습니다!
https://github.com/next-step/java-racingcar-kakao/pull/9/commits/4e84dd73a2af4f8851e02155c74f0bd36b421ab4

liquidjoo commented on src/test/java/carrace/CarMoveRuleTest.java:
```diff
@@ -0,0 +1,18 @@
+    @CsvSource({"0, false", "3, false", "4, true", "9, true"})
+    @ParameterizedTest
+    void move(int value, boolean move) {
+        CarMoveRule carMoveRule = new CarMoveRule(() -> value);
```
익명함수, 람다함수 활용 👍

liquidjoo reviewed (changes requested):
안녕하세요 영민님
이번 미션 리뷰어를 맡은 김성주 입니다 :)
미션 잘 구현해주셨어요 👍 
몇 가지 코멘트 남겼는데요~ 확인하고 다시 요청주세요 🙇

Gongmeda committed:
refactor: 가변 컬렉션 불변으로 반환 처리

Gongmeda committed:
refactor: static으로 Random 객체 생성 비용 절감 처리

Gongmeda committed:
refactor: static으로 Pattern 객체 생성 비용 절감 처리

Gongmeda committed:
refactor: 명시적으로 상수 분리

Gongmeda committed:
refactor: 자동차 이름 및 자동차 이름 중복 validation 로직 뷰에서 도메인 모델로 이동

Gongmeda committed:
refactor: round validation 로직 controller로 이동

Gongmeda committed:
refactor: StringCalculator에서 문자열 파싱 로직 분리

Gongmeda committed:
refactor: 테스트코드 @DisplayName 적용

Gongmeda committed:
feat: Round 클래스 구현

Gongmeda committed:
refactor: 전반적인 리팩토링

- Cars, Round 클래스 추가
- 하드 코딩괸 상수들 변수로 추출
- CarMoveRule 인터페이스로 추상화
- 깨지는 테스트 코드 수정

liquidjoo commented on src/main/java/carrace/CarRaceController.java:
```diff
@@ -17,37 +14,15 @@ public CarRaceController(CarRaceView view) {
-		proceedRounds(carRace, round);
-		endRace(carRace);
+		CarRace carRace = new CarRace(new DefaultCarMoveRule(new RandomNumberGenerator()), createCars(view.getCarNames()));
+		Round round = new Round(view.getCarRaceRound());
+		RaceResult raceResult = carRace.race(round);
+		view.displayRoundResult(raceResult);
```
피드백 반영 👍

liquidjoo commented on src/main/java/carrace/domain/Car.java:
```diff
@@ -1,12 +1,33 @@
+    private Car(String name, int position) {
         this.name = name;
+        this.position = position;
+        validateName();
```
```suggestion
        validateName(name, position);
        this.name = name;
        this.position = position;
```

빠른 반환을 통해서 유효성 검증하는 것도 좋은 방법 입니다 :)
조금이지만 변수에 할당되는 메모리의 이점이 있습니다

liquidjoo commented on src/main/java/carrace/domain/Car.java:
```diff
@@ -17,7 +38,20 @@ public int getPosition() {
 
-    public void moveForward() {
-        position++;
+    public Car moveForward() {
+        return new Car(name, position + MOVE_AMOUNT);
+    }
```
불변 객체의 활용 👍 
불변 객체의 이점에 대해서 따로 공부해보시면 재밌는 이야기가 많이 있습니다

liquidjoo commented on src/main/java/carrace/domain/CarMoveRule.java:
```diff
@@ -1,16 +1,5 @@
-        return numberGenerator.generate() >= MOVE_THRESHOLD;
-    }
+public interface CarMoveRule {
+    boolean isMovable();
```
인터페이스 활용 👍 
함수형 인터페이스가 될 수도 있으며 객체지향적설계의 이점을 배울 수 있을 것 같네요~

liquidjoo commented on src/main/java/carrace/domain/Cars.java:
```diff
@@ -0,0 +1,53 @@
+
+    public Cars(List<Car> cars) {
+        this.cars = cars;
+        validateDuplicateNames();
```
유효성 체크가 먼저되면 좋습니다 :)

liquidjoo commented on src/main/java/carrace/domain/Cars.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public List<Car> getCars() {
+        return cars;
```
방어적 복사에 대해서 알아보면 좋아요~

liquidjoo commented on src/main/java/carrace/domain/Cars.java:
```diff
@@ -0,0 +1,53 @@
+        return cars;
+    }
+
+    public Cars getMaxPositionCars() {
+        int maxPosition = cars.stream()
+                .mapToInt(Car::getPosition)
+                .max()
+                .orElseThrow();
+        return new Cars(cars.stream()
+                .filter(car -> car.getPosition() == maxPosition)
+                .collect(Collectors.toUnmodifiableList()));
+    }
```
해당 책임은 마지막 상태 값인 Cars 만 있으면 되지 않을까요?
RaceResult 으로 책임을 옮겨보는 것도 좋아 보이네요 :)

liquidjoo commented on src/test/java/carrace/domain/CarsTest.java:
```diff
@@ -0,0 +1,92 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatExceptionOfType;
+import static org.assertj.core.api.SoftAssertions.assertSoftly;
+
+class CarsTest {
```
꼼꼼한 테스트 코드 👍

liquidjoo reviewed (approved):
안녕하세요 영민님 :)
피드백 반영 잘해주셨어요 💯 
몇 가지 코멘트 남겼습니다~ 다음 단계에 반영해 보면 좋을 것 같네요
다음 단계 진행해 주세요 🙇
